cli/src/error.rs

```diff
@@ -0,0 +1,72 @@
+use nickel_lang_core::{eval::cache::lazy::CBNCache, program::Program};
+
+pub enum Error {
+    Program {
+        program: Program<CBNCache>,
+        error: nickel_lang_core::error::Error,
+    },
+    Io {
+        error: std::io::Error,
+    },
+    #[cfg(feature = "repl")]
+    Repl {
+        error: nickel_lang_core::repl::InitError,
+    },
+    #[cfg(feature = "format")]
+    Format {
+        error: crate::format::FormatError,
+    },
+}
+
+pub type CliResult<T> = Result<T, Error>;
+
+impl From<std::io::Error> for Error {
+    fn from(error: std::io::Error) -> Self {
+        Error::Io { error }
+    }
+}
+
+#[cfg(feature = "format")]
+impl From<crate::format::FormatError> for Error {
+    fn from(error: crate::format::FormatError) -> Self {
+        Error::Format { error }
+    }
+}
+
+#[cfg(feature = "repl")]
+impl From<nickel_lang_core::repl::InitError> for Error {
+    fn from(error: nickel_lang_core::repl::InitError) -> Self {
+        Error::Repl { error }
+    }
+}
+
+pub trait WithProgram<T> {
+    fn with_program(self, program: Program<CBNCache>) -> CliResult<T>;
+}
+
+impl<T> WithProgram<T> for Result<T, nickel_lang_core::error::Error> {
```

Comment:
I wonder if the naming is correct, as this is clearly an extension trait (cf https://rust-lang.github.io/rfcs/0445-extension-trait-conventions.html). Given the linked RFC, this might probably be called `ResultErrorExt` or maybe `CliResultExt`, even if `CliResult` is defined within this crate...or, `ResultErrorExt`? The latter is probably the closest to the RFC, but it's really obscure.

Also, I feel like the `T` should be a parameter of `with_program` and not of `WithProgram`, but it's purely aesthetic, as I believe the two version are isomorphic.

Another solution is to make `with_program` a free-standing function, which is perhaps less surprising but you lose the ability to chain with `.`. Or to simply inline `with_program`, which doesn't sound too bad either.

Reply:
> Making T just a parameter of the function doesn't work, I think, because I'm implementing WithProgram for Result<T, _> and the type parameter can't actually change for each invocation of with_program. That is to say, this would really be trait for things of kind Type -> Type. But that can't actually be expressed in Rust traits at the moment, as far as I know.

Ah, good point.

> My idea here was to get as close to just a ? at the use site as possible. Just implementing an instance of From isn't good enough here because I need to provide the Program that is to be used for reporting the error. Inlining with_program ends up exposing the structure of the error type at each use site, which feels a bit wrong to me.

I see your point, but I would argue that it's an extension trait currently because your return type is `CliResult<T>` and it seems to me that the only obvious implementer is `CliResult<T>`. I don't really see what other type would or could implement it, excepted maybe something that converts to `CliResult<T>`. Put differently, had we defined `CliResult` as a newtype wrapper, I believe `with_program` would naturally be a method of `CliResult` rather than having its own trait.

As per the RFC:

> Of course, adding methods via a new trait happens all the time. What makes it an extension trait is that the trait is not designed for generic use, but only as way of adding methods to a specific type or family of types.
>
>This is of course a somewhat subjective distinction. Whenever designing an extension trait, one should consider whether the trait could be used in some more generic way. If so, the trait should be named and exported as if it were just a "normal" trait. But traits offering groups of methods that really only make sense in the context of some particular type(s) are true extension traits.

But, as mentioned in this extract, this is a subjective distinction. All in all this is not very important and I'll let you decide what you think is best. I just wanted to bring the attention to existing conventions.

Reply:
It's not `CliResult<T>` that implements the trait but `Result<T, nickel_lang_core::error::Error>`. Note the different `Error` type. In a way, it's really a specialized `Into` that needs to take an extra parameter. But I do see the point about it maybe being an extension trait per the RFC now. And I really don't see how this particular trait could be generically useful 😅

So, I agree with renaming it to something like `ResultErrorExt`. Naming things is truly one of the hard problems :laughing: